File: backend/app/schemas/analytics.py
from pydantic import BaseModel


class ClientCount(BaseModel):
    client: str
    count: int


class LevelCount(BaseModel):
    level: str
    count: int


class MonthlyTrend(BaseModel):
    month: str  # "2025-01"
    new_hires: int
    cumulative: int


class SpanOfControl(BaseModel):
    manager_id: str
    manager_name: str
    designation: str
    direct_report_count: int
    is_outlier: bool = False  # flagged if < 2 or > 10


class CrossReport(BaseModel):
    employee_id: str
    employee_name: str
    employee_designation: str
    external_manager_id: str
    external_manager_name: str
    external_manager_location: str
    relationship_type: str


class ProjectSummary(BaseModel):
    id: str
    name: str
    status: str
    member_count: int
    client_name: str


class BranchAnalytics(BaseModel):
    total_headcount: int
    active_count: int
    client_breakdown: list[ClientCount]
    level_breakdown: list[LevelCount]
    monthly_trend: list[MonthlyTrend]
    span_of_control: list[SpanOfControl]
    hierarchy_depth: int
    departments_without_manager: list[str]
    cross_reports: list[CrossReport]
    projects: list[ProjectSummary]
    orphaned_projects: list[ProjectSummary]

File: frontend/src/types/analytics.ts
export interface ClientCount {
  client: string
  count: number
}

export interface LevelCount {
  level: string
  count: number
}

export interface MonthlyTrend {
  month: string
  new_hires: number
  cumulative: number
}

export interface SpanOfControl {
  manager_id: string
  manager_name: string
  designation: string
  direct_report_count: number
  is_outlier: boolean
}

export interface CrossReport {
  employee_id: string
  employee_name: string
  employee_designation: string
  external_manager_id: string
  external_manager_name: string
  external_manager_location: string
  relationship_type: string
}

export interface ProjectSummary {
  id: string
  name: string
  status: string
  member_count: number
  client_name: string
}

export interface BranchAnalytics {
  period?: string | null
  data_source?: "hrms" | "excel"
  total_headcount: number
  active_count: number
  client_breakdown: ClientCount[]
  level_breakdown: LevelCount[]
  monthly_trend: MonthlyTrend[]
  span_of_control: SpanOfControl[]
  hierarchy_depth: number
  departments_without_manager: string[]
  cross_reports: CrossReport[]
  projects: ProjectSummary[]
  orphaned_projects: ProjectSummary[]
}
